File: algorithm/python/sorting/QuickSort.py
# -*-encoding:utf-8-*-
# QuickSort.py
# cww97
# 2017/09/26
from tracer import *
from random import randint
n = 10

# define tracers
chart = ChartTracer('Chart')
tracer = Array1DTracer('Array').attach(chart.getID()) # debugby wyj
logger = LogTracer('Output')

# init data
arr = []
logger.Print('original array:')
for i in range(n):
    arr.append(randint(0, 20))
    logger.Print(str(arr[i]))
tracer.setData(arr)

def qsort(l, r):
    i = l
    j = r
    mid = arr[(l + r) >> 1]
    while i <= j:
        tracer.select(l).select(r).wait(1)  # -1
        while arr[i] < mid:
            i += 1
            tracer.select(i).wait(1).deselect(i)  # -1
        while mid < arr[j]:
            j -= 1
            tracer.select(j).wait(1).deselect(j)  # -1
        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            tracer.notify(i, arr[i]).notify(j, arr[j]).wait(1)  # -1
            tracer.denotify(i).denotify(j)
            i += 1
            j -= 1
    if i < r:
        qsort(i, r)
    if l < j:
        qsort(l, j)

if __name__ == '__main__':
    qsort(0, n - 1)
    logger.Print('sorted array:')
    for i in range(n):
        logger.Print(str(arr[i]))
    tracerlist.info()

File: algorithm/c++/quick_sort.cpp
#include "jsoncppheader.h"
#include <string>
int D[100] = {9,8,8,6,1,7,5,7,11,9,4,2};
int N = 12;
ChartTracer chart = ChartTracer("Chart");
Array1DTracer tracer = Array1DTracer("Quicksort").setData(D, N);
LogTracer logger = LogTracer();
void partition(int* D,int low,int high) {
    int i, j, s;
    while (high > low) {
        i = low;
        j = high;
        s = D[low];
        while (i < j) {
            tracer.select(high).select(low).wait();
            while (D[j] > s){ 
                tracer.select(j).wait();
                tracer.deselect(j);
                j--;
            }
            D[i] = D[j];
            tracer.notify(i, D[j]).wait().denotify(i);
            while (s >= D[i] && i < j){ 
                tracer.select(i).wait();
                tracer.deselect(i);
                i++;
            }
            D[j] = D[i];
            tracer.notify(j, D[i]).wait().denotify(j);
            tracer.deselect(high).deselect(low);
        }
        D[i] = s;
        tracer.notify(i, s).wait();
        tracer.denotify(i);
        partition(D, low, i-1);
        low = i+1;
    }
}

void quicksort(int* D) {
    partition(D, 0, N-1);
}
int main()
{
    tracer.attach(chart.id);
	string s = "original array = [";
	for(int i = 0; i < N - 1; i++)
		s+=to_string(D[i])+",";
	s+=to_string(D[N - 1])+"]";
	logger.print(s);
	quicksort(D);
	s = "sorted array = [";
	for(int i = 0; i < N - 1; i++)
		s+=to_string(D[i])+",";
	s+=to_string(D[N - 1])+"]";
	logger.print(s);
}